src/Symfony/Component/Mailer/Transport.php

```diff
@@ -0,0 +1,180 @@
+use Symfony\Component\Mailer\Bridge\Postmark;
+use Symfony\Component\Mailer\Bridge\Sendgrid;
+use Symfony\Component\Mailer\Exception\InvalidArgumentException;
+use Symfony\Component\Mailer\Exception\LogicException;
+use Symfony\Component\Mailer\Transport\TransportInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @experimental in 4.3
+ */
+class Transport
+{
+    public static function fromDsn(string $dsn, EventDispatcherInterface $dispatcher = null, HttpClientInterface $client = null, LoggerInterface $logger = null): TransportInterface
+    {
+        // failover?
+        $dsns = preg_split('/\s++\|\|\s++/', $dsn);
+        if (\count($dsns) > 1) {
+            $transports = [];
+            foreach ($dsns as $dsn) {
+                $transports[] = self::createTransport($dsn, $dispatcher, $client, $logger);
+            }
+
+            return new Transport\FailoverTransport($transports);
+        }
+
+        // round robin?
+        $dsns = preg_split('/\s++&&\s++/', $dsn);
+        if (\count($dsns) > 1) {
+            $transports = [];
+            foreach ($dsns as $dsn) {
+                $transports[] = self::createTransport($dsn, $dispatcher, $client, $logger);
+            }
+
+            return new Transport\RoundRobinTransport($transports);
+        }
+
+        return self::createTransport($dsn, $dispatcher, $client, $logger);
+    }
+
+    private static function createTransport(string $dsn, EventDispatcherInterface $dispatcher = null, HttpClientInterface $client = null, LoggerInterface $logger = null): TransportInterface
+    {
+        if (false === $parsedDsn = parse_url($dsn)) {
+            throw new InvalidArgumentException(sprintf('The "%s" mailer DSN is invalid.', $dsn));
+        }
+
+        if (!isset($parsedDsn['host'])) {
+            throw new InvalidArgumentException(sprintf('The "%s" mailer DSN must contain a mailer name.', $dsn));
+        }
+
+        $user = \urldecode($parsedDsn['user'] ?? '');
+        $pass = \urldecode($parsedDsn['pass'] ?? '');
+        \parse_str($parsedDsn['query'] ?? '', $query);
+
+        switch ($parsedDsn['host']) {
+            case 'null':
+                if ('smtp' === $parsedDsn['scheme']) {
+                    return new Transport\NullTransport($dispatcher, $logger);
+                }
+
+                throw new LogicException(sprintf('The "%s" scheme is not supported for mailer "%s".', $parsedDsn['scheme'], $parsedDsn['host']));
+            case 'sendmail':
+                if ('smtp' === $parsedDsn['scheme']) {
+                    return new Transport\SendmailTransport(null, $dispatcher, $logger);
+                }
+
+                throw new LogicException(sprintf('The "%s" scheme is not supported for mailer "%s".', $parsedDsn['scheme'], $parsedDsn['host']));
+            case 'gmail':
+                if (!class_exists(Google\Smtp\GmailTransport::class)) {
```

Comment:
@fabpot I don't understand how this class would not exist. The bridges are inside the component. Also why do the bridges have `composer.json` ?

Reply:
The providers are going to be subtree-split-ed.